services/QuillLMS/app/models/feedback_history.rb

```diff
@@ -94,11 +95,60 @@ def serializable_hash(options = nil)
+  end
+
+  def serialize_by_activity_session_detail
+   serializable_hash(only: [:entry, :feedback_text, :feedback_type, :optimal, :used], include: []).symbolize_keys
+  end
+
   def self.batch_create(param_array)
     param_array.map { |params| create(params) }
   end
 
   private def confirm_prompt_type
     self.prompt_type = DEFAULT_PROMPT_TYPE if prompt_id && !prompt_type
   end
+
+  def self.list_by_activity_session(activity_id: nil, page: 1, page_size: DEFAULT_PAGE_SIZE)
+    query = select(%{
+        feedback_histories.activity_session_uid AS session_uid,
+        MIN(feedback_histories.time) AS start_date,
+        comprehension_prompts.activity_id,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' THEN 1 END) AS because_attempts,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' THEN 1 END) AS but_attempts,
+        COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' THEN 1 END) AS so_attempts,
+        (
+          ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' AND feedback_histories.optimal THEN 1 END) = 1) OR
+            (COUNT(CASE WHEN comprehension_prompts.conjunction = 'because' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'because' THEN comprehension_prompts.max_attempts END))) AND
+          ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' AND feedback_histories.optimal THEN 1 END) = 1) OR
+            (COUNT(CASE WHEN comprehension_prompts.conjunction = 'but' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'but' THEN comprehension_prompts.max_attempts END))) AND
+          ((COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' AND feedback_histories.optimal THEN 1 END) = 1) OR
+            (COUNT(CASE WHEN comprehension_prompts.conjunction = 'so' THEN 1 END) = MAX(CASE WHEN comprehension_prompts.conjunction = 'so' THEN comprehension_prompts.max_attempts END)))
+        ) AS complete
+      })
+      .joins("LEFT OUTER JOIN comprehension_prompts ON feedback_histories.prompt_id = comprehension_prompts.id")
+      .where(used: true)
+      .group(:activity_session_uid, :activity_id)
+      .order('start_date DESC')
+    query = query.where(comprehension_prompts: {activity_id: activity_id.to_i}) if activity_id
+    query = query.limit(page_size)
+    query = query.offset((page.to_i - 1) * page_size.to_i) if page && page.to_i > 1
+    query
+  end
+
+  def self.serialize_detail_by_activity_session(activity_session_uid)
+    history = FeedbackHistory.list_by_activity_session.where(activity_session_uid: activity_session_uid).first
+    return nil unless history
+    histories = FeedbackHistory.where(activity_session_uid: activity_session_uid).all
+
+    output = history.serialize_by_activity_session
+    prompt_groups = histories.group_by { |h| h&.prompt&.conjunction }.map { |k,v| [k, {prompt_id: v.first.prompt_id, attempts: v}] }.to_h.symbolize_keys
```

Comment:
This would be more parsable for future developers if you used descriptive names for the local vars instead of `|k,v|`, e.g. `|conjunction, attempts|`. This also is complex enough it might be worth using several multi-line`do || end` blocks instead the inline.

Reply:
Yeah, that makes sense.  I think I went overboard in correcting toward "I bet we can do this in one-liners".  Revised.